### PR TITLE
Modify tip cardano-cli query to also return the current epoch

### LINKED ISSUE #1737: [FR] - Add epoch and slot in epoch to tip command
**Internal/Exernal**
*Internal*   if an IOHK staff member.

**Describe the feature you'd like**
A clear and detailed description of what you are suggesting.

**Describe alternatives you've considered**

**Additional context / screenshots**

### PR DESCRIPTION
**Pull request by Jimbo4350**
Closes #1737 
```
> cabal exec cardano-cli -- query tip --shelley-era --shelley-mode --testnet-magic 42
{
    "epoch": 0,
    "hash": "19fdfc54e7a1e2fb7ff45254a6aa0aa88ea7a80de4b48297924c9abf353f9cb7",
    "slot": 42,
    "block": 3
}
```

### CODE REVIEW

File: cardano-api/src/Cardano/Api/Block.hs [hunk 1]
```diff
@@ -250,7 +250,7 @@ getBlockHeader (ByronBlock block)
 --
 
 data ChainPoint = ChainPointAtGenesis
-                | ChainPoint !SlotNo !(Hash BlockHeader)
+                | ChainPoint !SlotNo !(Hash BlockHeader) (Maybe EpochNo)
```
Comment (dcoutts) [inline cardano-api/src/Cardano/Api/Block.hs:253]: Oh, no we cannot change this. A point is fundamental. This is what it is, no more no less.

In the CLI we can pair up a point with an epoch number, but we cannot change the point type.

**Review by dcoutts — CHANGES_REQUESTED** [review]

**Review by dcoutts — CHANGES_REQUESTED** [review]
I think this will be better to do entirely in the CLI in the short term. I don't think we want to change the `getLocalChainTip`, since it changes it from being a single chain sync query to being two protocols, with a possibility of failure. It doesn't make for a nice API.

In the CLI we can use the existing `getLocalChainTip`, and then use the existing `queryNodeLocalState` with the `QueryEpoch` query. That would not need any API changes.

Now it's true that this is imperfect, since it needs two queries that are not guaranteed to be synchronised (which is the root of the "Maybe" problem in the API in the current version of the patch).

What would be better here, in the medium term, would be to properly make use of the ability of the query to supply no point and implicitly acquire the current point. Then we can use `QueryChainPoint` and `QueryEpoch`. The only thing we're missing to make this perfect is `QueryChainBlockNo`. If we add that query to the consensus, and expose it in the API, then we can do this all as a single query (or rather a single atomic batch of queries).

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs [hunk 2]
```diff
@@ -147,21 +148,47 @@ writeProtocolParameters mOutFile pparams =
...
+              eResult <- liftIO $ queryNodeLocalState localNodeConnInfo cPoint epochQuery
+              case eResult of
+                Left _acqFail -> return Nothing
+                Right eNum -> case eNum of
+                                Left _eraMismatch -> return Nothing
+                                Right epochNum -> return $ Just epochNum
+
+    let pTip = encodePretty tip
+        epoch = encodePretty $ toObject mEpoch
+        output = LBS.concat [pTip, LBS.fromStrict (Text.encodeUtf8 "\n"), epoch]
```
Comment (dcoutts) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs:181]: How about merging these into one json object?
```
output = encodePretty (tip <> toObject mEpoch)
```

One object as output is probably what people want anyway.

**Review by dcoutts — APPROVED** [review]
This looks much better.

Two minor comments below.

Lets squash down to one patch again, and clean up the unnecessary changes in whitespace and comments (that are just there as a result of us going back and forth on things).

**Review by olgahryniuk — APPROVED** [review]

**Pull request merged**